File: src/api/sse/daemons.py
import asyncio
import os
import time

import orjson
import psutil

from api.core.health import (
    _evaluate_cache_health,
    _evaluate_database_health,
    _evaluate_storage_health,
)
from api.core.metrics import _metrics, _start_time
from config.provider import GlobalConfigProvider

from .connection_manager import sse_mgr


async def health_poller() -> None:
    """
    Background daemon: Evaluates system health periodically and pushes to SSE.
    Reuses existing health check logic safely.
    """
    config = GlobalConfigProvider().get_config()
    interval = config.sse.health_interval

    while True:
        try:
            db_status, db_up = await _evaluate_database_health(config)
            cache_status = await _evaluate_cache_health(config)
            storage_status = _evaluate_storage_health(config)
            proc = psutil.Process(os.getpid())

            all_up = db_up and (cache_status.get("status") == "up")

            payload = orjson.dumps(
                {
                    "status": "healthy" if all_up else "degraded",
                    "checks": {
                        "databases": db_status,
                        "cache": cache_status,
                        "storages": storage_status,
                    },
                    "system": {
                        "memory_mb": round(proc.memory_info().rss / 1048576, 2),
                        "cpu_pct": proc.cpu_percent(interval=None),
                    },
                }
            ).decode()

            await sse_mgr.publish("system:health", "health", payload)
        except Exception:
            pass  # Best effort, never crash the daemon

        await asyncio.sleep(interval)


async def metrics_pusher() -> None:
    """
    Background daemon: Periodically pushes live metrics snapshot.
    """
    config = GlobalConfigProvider().get_config()
    interval = config.sse.metrics_interval

    while True:
        try:
            proc = psutil.Process(os.getpid())
            payload = orjson.dumps(
                {
                    "uptime_seconds": round(time.time() - _start_time, 2),
                    "memory_mb": round(proc.memory_info().rss / 1048576, 2),
                    "cpu_percent": proc.cpu_percent(interval=None),
                    "db_queries_total": _metrics.get("db_queries_total", 0),
                    "sse_connections": sse_mgr.active_count,
                    "sse_topics": sse_mgr.topic_count,
                }
            ).decode()

            await sse_mgr.publish("metrics", "metrics", payload)
        except Exception:
            pass

        await asyncio.sleep(interval)

File: src/api/sse/daemons.rs
use serde_json::json;
use std::time::Duration;
use tokio::time::sleep;

use crate::api::sse::connection_manager::SSE_MGR;
use crate::config::loader::ConfigManager;

fn get_system_stats() -> (f64, f64) {
    // Lightweight system stats without pulling in sysinfo crate
    // Returns (cpu_pct, memory_mb) - approximate values
    (0.0, 0.0) // Placeholder - extend with sysinfo when needed
}

fn get_uptime() -> u64 {
    // Simple uptime tracking
    use std::time::{SystemTime, UNIX_EPOCH};
    static START_TIME: std::sync::OnceLock<u64> = std::sync::OnceLock::new();
    let start = *START_TIME.get_or_init(|| {
        SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs()
    });
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs()
        .saturating_sub(start)
}

pub async fn health_poller() {
    let config = ConfigManager::get();
    let interval = Duration::from_secs(config.sse.health_interval as u64);

    loop {
        let (cpu_pct, mem_mb) = get_system_stats();
        let payload = json!({
            "status": "healthy",
            "checks": {
                "databases": "up",
                "cache": "up",
                "storages": "up"
            },
            "system": {
                "memory_mb": mem_mb,
                "cpu_pct": cpu_pct,
                "uptime_seconds": get_uptime()
            }
        })
        .to_string();

        SSE_MGR.publish("system:health", "health", payload).await;

        let count = SSE_MGR.active_count().await;
        if count > 0 {
            // Cleanup handled implicitly in publish when send fails
        }

        sleep(interval).await;
    }
}

pub async fn metrics_pusher() {
    let config = ConfigManager::get();
    let interval = Duration::from_secs(config.sse.metrics_interval as u64);

    loop {
        let active_count = SSE_MGR.active_count().await;
        let topic_count = SSE_MGR.topic_count().await;
        let (cpu_pct, mem_mb) = get_system_stats();

        let payload = json!({
            "uptime_seconds": get_uptime(),
            "memory_mb": mem_mb,
            "cpu_percent": cpu_pct,
            "db_queries_total": 0,
            "sse_connections": active_count,
            "sse_topics": topic_count
        })
        .to_string();

        SSE_MGR.publish("metrics", "metrics", payload).await;

        sleep(interval).await;
    }
}
